handler/src/main/java/io/netty/handler/ssl/OpenSslSessionCache.java

```diff
@@ -0,0 +1,256 @@
+            } catch (SSLException e) {
+                // TODO: Should we log this ?
+                return false;
+            }
+            if (!sessionCreated(session)) {
+                return false;
+            }
+
+            final OpenSslSession old = sessions.put(session.sessionId(), session.retain());
+            if (old != null) {
+                sessionRemoved(old);
+                old.release();
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public final long getSession(long ssl, byte[] sessionId) {
+        OpenSslSessionId id = new OpenSslSessionId(sessionId);
+        synchronized (this) {
+            OpenSslSession session = sessions.get(id);
+            if (session == null) {
+                return -1;
+            }
+            long nativeAddr = session.nativeAddr();
+            if (nativeAddr == -1 || !session.isValid()) {
+                removeSession(session);
+                return -1;
+            }
+
+            // This needs to happen in the synchronized block so we ensure we never destroy it before we incremented
+            // the reference count.
+            if (!io.netty.internal.tcnative.SSLSession.upRef(nativeAddr)) {
+                // we could not increment the reference count, something is wrong. Let's just drop the session.
+                removeSession(session);
+                return -1;
+            }
+
+            session.retain();
+
+            if (io.netty.internal.tcnative.SSLSession.shouldBeSingleUse(nativeAddr)) {
+                // Should only be used once
+                removeSession(session);
+            }
+            session.updateLastAccessedTime();
+            return nativeAddr;
+        }
+    }
+
+    protected void removeSession(OpenSslSession session) {
+        sessions.remove(session.sessionId());
+        sessionRemoved(session);
+        session.release();
+    }
+
+    final SSLSession getSession(byte[] bytes) {
+        OpenSslSessionId id = new OpenSslSessionId(bytes);
+        synchronized (this) {
+            OpenSslSession session = sessions.get(id);
+            if (session == null) {
+                return null;
+            }
+            if (!session.isValid()) {
+                removeSession(session);
+                return null;
+            }
+            return session;
+        }
+    }
+
+    final synchronized List<byte[]> getIds() {
+        List<byte[]> ids = new ArrayList<byte[]>(sessions.size());
+        for (OpenSslSession session: sessions.values()) {
+            if (session.isValid()) {
+                ids.add(session.getId());
+            }
+        }
+        return ids;
+    }
+
+    final synchronized void freeSessions() {
+        final OpenSslSession[] sessionsArray = sessions.values().toArray(new OpenSslSession[0]);
+        sessions.clear();
+
+        for (OpenSslSession session: sessionsArray) {
+            sessionRemoved(session);
+            session.release();
```

Comment:
The `freeSeessions()` method does the same thing as `expungeInvalidSessions()` (except `synchronized`). Consider reusing the logic that clears `sessions`.

Reply:
it does not the same as `exponugeInvalidSessions` also takes `isValid` into account